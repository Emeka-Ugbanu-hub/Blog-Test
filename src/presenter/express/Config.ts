
import Service from '../../service/Service';
import { LoggerInstance } from 'winston';

export default interface Config {
  readonly morganLogFormat: string,
  readonly morganDirectory: string,
  readonly service: Service,
  readonly logger: LoggerInstance;
}