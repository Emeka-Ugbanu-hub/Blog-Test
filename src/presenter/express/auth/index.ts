import login from './login';
import register from './register';
import forgetPassword from './forgetPassword';

export {
  login, 
  register,
  forgetPassword
}