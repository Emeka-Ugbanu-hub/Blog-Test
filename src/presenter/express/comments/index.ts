import createComment from './createComment';
import getCommentById from './getCommentById';
export {
  createComment,
  getCommentById
}