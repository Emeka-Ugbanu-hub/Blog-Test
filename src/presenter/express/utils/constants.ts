const OK_200_HTTP_CODE = 200;
const CREATED_201_HTTP_CODE = 201;
const NO_CONTENT_204_HTTP_CODE = 204;
const BAD_REQUEST_400_HTTP_CODE = 400;
const UNAUTHORISED_401_HTTP_CODE = 401;
const FORBIDDEN_403_HTTP_CODE = 403;
const NOT_FOUND_404_HTTP_CODE = 404;
const CONFLICT_409_HTTP_CODE = 409;
const UNPROCESSABLE_ENTITY_422_HTTP_CODE = 422;
const SERVER_ERROR_500_HTTP_CODE = 500;


export {
  OK_200_HTTP_CODE,
  CREATED_201_HTTP_CODE,
  NO_CONTENT_204_HTTP_CODE,
  BAD_REQUEST_400_HTTP_CODE,
  UNAUTHORISED_401_HTTP_CODE,
  FORBIDDEN_403_HTTP_CODE,
  NOT_FOUND_404_HTTP_CODE,
  CONFLICT_409_HTTP_CODE,
  UNPROCESSABLE_ENTITY_422_HTTP_CODE,
  SERVER_ERROR_500_HTTP_CODE,
}