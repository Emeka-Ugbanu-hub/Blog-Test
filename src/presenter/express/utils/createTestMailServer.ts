const MailDev = require('maildev');
import config from '../../../config';

export default () => {
  return new Promise((resolve) => {
    const maildev = new MailDev({
      port : config.mail.port,
      ignoreTLS : true
    });
    maildev.close();
    maildev.listen();
    maildev.on('new', (email: any) => {
      resolve(email);
      maildev.close();
    });
  });
};