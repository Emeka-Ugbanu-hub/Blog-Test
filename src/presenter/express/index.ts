import { Router, Request, Response} from 'express';
import Config from './Config';
import enhancedRouter from './enhancedRouter';
import {register, login, forgetPassword,resetPassword} from './auth';
import {getPosts, createPost, getPostById, updatePost, deletePostById} from './posts';
import {getUserById, getUsers, updateUser, createUser, deleteUserById} from './users';
import {createComment, getCommentById, getComments, updateComment} from './comments';

export default (config: Config): Router => {
  const router: Router = enhancedRouter(config);

  router.get('/', (req: Request, res: Response) => {
    res.status(200)
       .json({message: "This is where the awesomeness happen..."});
  });

  router.post('/auth/register', register(config));
  router.post('/auth/login', login(config));
  router.post('/auth/forget-password', forgetPassword(config));
  router.post('/auth/reset-password', resetPassword(config));

  router.post('/users', createUser(config));
  router.delete('/users/:user_id', deleteUserById(config));
  router.get('/users/:user_id', getUserById(config));
  router.get('/users', getUsers(config));
  router.patch('/users/:user_id', updateUser(config));
  
  router.get('/posts', getPosts(config));
  router.get('/posts/:post_id', getPostById(config));
  router.delete('/posts/:post_id', deletePostById(config));
  router.post('/posts', createPost(config));
  router.patch('/posts/:post_id', updatePost(config));

  router.get('/comments', getComments(config));
  router.post('/comments', createComment(config));
  router.get('/comments/:comment_id', getCommentById(config));
  router.patch('/comments/:comment_id', updateComment(config));

  return router;
}