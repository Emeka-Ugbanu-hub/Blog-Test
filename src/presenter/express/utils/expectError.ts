import {UNPROCESSABLE_ENTITY_422_HTTP_CODE} from './constants';

interface Response {
  status: number;
  body: any;
}

export default (response: Response, expectedStatusCode: number = UNPROCESSABLE_ENTITY_422_HTTP_CODE) => {
  expect(response.status).toBe(expectedStatusCode);
  expect(response.body).toMatchSnapshot();
}