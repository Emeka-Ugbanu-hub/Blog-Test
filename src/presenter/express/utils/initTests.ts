import config from '../../../config';
import * as sourceMapSupport from 'source-map-support';
sourceMapSupport.install();
import * as iconvLite from 'iconv-lite';
iconvLite.encodingExists('foo');
import * as express from 'express';
import * as createSupertest from 'supertest';
import logger from '../../../logger';
import presenter from '../index';
import serviceFactory from '../../../service/factory';
import {API_ROUTE_V1} from '../../../utils/constants';

const app: express.Application = express();

const service = serviceFactory();

const presenterFacade = presenter({
  morganLogFormat: config.express.morganLogFormat,
  morganDirectory: config.express.morganDirectory,
  service,
  logger
});

app.use(API_ROUTE_V1, presenterFacade);

const request = createSupertest(app);

export default () => {

  beforeAll(async() => {
    await service.migrate();
  });

  beforeEach(async() => {
    await service.migrate();
  });
  
  afterEach(async() => {
    await service.rollback();
  });

  afterAll(async() => {
    await service.clearService();
  });
  
  return { service, request };
};