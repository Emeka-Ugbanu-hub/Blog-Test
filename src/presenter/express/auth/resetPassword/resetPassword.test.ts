import initTests from '../../utils/initTests';
import {API_ROUTE_V1, DEFAULT_USER_PERMISSIONS} from '../../../../utils/constants';
import {Response} from 'express';
import {CREATED_201_HTTP_CODE, OK_200_HTTP_CODE, UNPROCESSABLE_ENTITY_422_HTTP_CODE, CONFLICT_409_HTTP_CODE} from '../../utils/constants';
import config from '../../../../config';
import {TEST_VALID_LOGIN_USER, TEST_INVALID_EMAIL,TEST_VALID_PASSWORD, TEST_VALID_EMAIL, 
  TEST_TOO_SHORT_PASSWORD,TEST_NOT_MATCHING_RESET_PASSWORD_TOKEN , TEST_DIFFERENT_VALID_PASSWORD, TEST_INVALID_RESET_PASSWORD_TOKEN} from '../../../../utils/testValues';
import * as R  from 'ramda';
import expectError from '../../utils/expectError';
import * as moment from 'moment';
import createTestMailServer from '../../utils/createTestMailServer';
import generateRandomToken from '../../../../utils/generateRandomToken';
import verifyPassword from '../../../../utils/verifyPassword';

const { service, request } = initTests();

describe(__filename, () => {

  it('should fail to reset password without input', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/reset-password`);
    expectError(response);
  });

  it('should fail to reset password without token', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/reset-password`)
                                  .send({
                                    password: TEST_VALID_PASSWORD,
                                    password_confirmation: TEST_VALID_PASSWORD
                                  });
    expectError(response);
  });

  it('should fail to reset password without password', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/reset-password`)
                                  .send({
                                    token: TEST_INVALID_RESET_PASSWORD_TOKEN,
                                    password_confirmation: TEST_VALID_PASSWORD
                                  });
    expectError(response);
  });

  it('should fail to reset password without password_confirmation', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/reset-password`)
                                  .send({
                                    token: TEST_INVALID_RESET_PASSWORD_TOKEN,
                                    password: TEST_VALID_PASSWORD
                                  });
    expectError(response);
  });

  it('should fail to reset password without passwords match', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/reset-password`)
                                  .send({
                                    token: TEST_INVALID_RESET_PASSWORD_TOKEN,
                                    password: TEST_VALID_PASSWORD,
                                    password_confirmation: TEST_DIFFERENT_VALID_PASSWORD
                                  });
    expectError(response);
  });

  it('should fail to reset password with too short passwords', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/reset-password`)
                                  .send({
                                    token: TEST_INVALID_RESET_PASSWORD_TOKEN,
                                    password: TEST_TOO_SHORT_PASSWORD,
                                    password_confirmation: TEST_TOO_SHORT_PASSWORD
                                  });
    expectError(response);
  });

  it('should fail to reset password when there is no match', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/reset-password`)
                                  .send({
                                    token: TEST_INVALID_RESET_PASSWORD_TOKEN,
                                    password: TEST_VALID_PASSWORD,
                                    password_confirmation: TEST_DIFFERENT_VALID_PASSWORD
                                  });
    expectError(response);
  });

  it('should fail to reset password when there is no matching token in db', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/reset-password`)
                                  .send({
                                    token: TEST_NOT_MATCHING_RESET_PASSWORD_TOKEN,
                                    password: TEST_VALID_PASSWORD,
                                    password_confirmation: TEST_VALID_PASSWORD
                                  });
    expectError(response);
  });

  it('should fail to reset password when the token is expired', async () => {
    
    const registeredUser = await service.register({
      email: TEST_VALID_EMAIL, 
      password: TEST_VALID_PASSWORD
    });
    const token = generateRandomToken();
    await service.createResetPasswordToken({
      userId: registeredUser.user.id, 
      token, 
      createdAt: moment().subtract(2, 'h').toString()
    });

    const response = await request.post(`${API_ROUTE_V1}/auth/reset-password`)
                                  .send({
                                    token,
                                    password: TEST_VALID_PASSWORD,
                                    password_confirmation: TEST_VALID_PASSWORD
                                  });
    expectError(response);
  });

  it('should successfuly change user password, delete token, and send email', async () => {
    const mailServer = createTestMailServer();
    const registeredUser = await service.register({
      email: TEST_VALID_EMAIL, 
      password: TEST_VALID_PASSWORD
    });

    const token = generateRandomToken();

    await service.createResetPasswordToken({
      userId: registeredUser.user.id, 
      token, 
      createdAt: moment().subtract(59, 'minutes').toString()
    });

    const response = await request.post(`${API_ROUTE_V1}/auth/reset-password`)
                                  .send({
                                    token,
                                    password: TEST_DIFFERENT_VALID_PASSWORD,
                                    password_confirmation: TEST_DIFFERENT_VALID_PASSWORD
                                  });

    const updatedUser: any = await service.getUserById({id: registeredUser.user.id});                          
    const tokenAfterReseting = await service.getResetPasswordTokenByToken({ token });
    const match = await verifyPassword(TEST_DIFFERENT_VALID_PASSWORD, updatedUser.password);
    const email: any = await mailServer;

    expect(email.from[0].address).toBe(config.mail.from);
    expect(email.to[0].address).toBe(registeredUser.user.email);
    expect(match).toBe(true);
    expect(tokenAfterReseting).toBeNull();
    expect(response.status).toBe(OK_200_HTTP_CODE);
    expect(response.body).toMatchSnapshot();
  });
});