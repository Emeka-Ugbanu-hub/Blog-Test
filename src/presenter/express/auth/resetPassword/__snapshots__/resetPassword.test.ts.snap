// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/resetPassword/resetPassword.test.ts should fail to reset password when the token is expired 1`] = `
Object {
  "message": "Expired reset password token",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/resetPassword/resetPassword.test.ts should fail to reset password when the token is expired 2`] = `
Object {
  "message": "Your password has been changed sucessfully. You can not log in.",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/resetPassword/resetPassword.test.ts should fail to reset password when there is no match 1`] = `
Object {
  "errors": Array [
    "Passwords must match in user",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/resetPassword/resetPassword.test.ts should fail to reset password when there is no matching token in db 1`] = `
Object {
  "message": "Invalid reset password token",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/resetPassword/resetPassword.test.ts should fail to reset password with too short passwords 1`] = `
Object {
  "errors": Array [
    "Required data in user.password must have at least 6 characters.",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/resetPassword/resetPassword.test.ts should fail to reset password without input 1`] = `
Object {
  "errors": Array [
    "Missing required value in 'user.token'",
    "Missing required value in 'user.password'",
    "Missing required value in 'user.password_confirmation'",
    "Passwords must match in user",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/resetPassword/resetPassword.test.ts should fail to reset password without password 1`] = `
Object {
  "errors": Array [
    "Missing required value in 'user.password'",
    "Passwords must match in user",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/resetPassword/resetPassword.test.ts should fail to reset password without password_confirmation 1`] = `
Object {
  "errors": Array [
    "Missing required value in 'user.password_confirmation'",
    "Passwords must match in user",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/resetPassword/resetPassword.test.ts should fail to reset password without passwords match 1`] = `
Object {
  "errors": Array [
    "Passwords must match in user",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/resetPassword/resetPassword.test.ts should fail to reset password without token 1`] = `
Object {
  "errors": Array [
    "Missing required value in 'user.token'",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/resetPassword/resetPassword.test.ts should successfuly change user password, delete token, and send email 1`] = `
Object {
  "message": "Your password has been changed sucessfully. You can not log in.",
}
`;
